include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -2098,96 +2014,76 @@ namespace internal
+
+    using Eval =
+      EvaluatorTensorProduct<symmetric_evaluate ? internal::evaluate_evenodd :
+                                                  internal::evaluate_general,
+                             dim - 1,
+                             fe_degree + 1,
+                             n_q_points_1d,
+                             Number>;
+
+    using EvaluationData =
+      FEEvaluationBaseData<dim, typename Number::value_type, true, Number>;
+
+    static Eval
+    create_evaluator_tensor_product(
+      const internal::MatrixFreeFunctions::UnivariateShapeData<Number> &data,
+      const unsigned int subface_index,
+      const unsigned int direction)
+    {
+      if (symmetric_evaluate)
+        return Eval(data.shape_values_eo,
+                    data.shape_gradients_eo,
+                    data.shape_hessians_eo,
+                    data.fe_degree + 1,
+                    data.n_q_points_1d);
+      else if (subface_index >= GeometryInfo<dim>::max_children_per_cell)
+        return Eval(data.shape_values,
+                    data.shape_gradients,
+                    data.shape_hessians,
+                    data.fe_degree + 1,
+                    data.n_q_points_1d);
+      else
+        {
+          const unsigned int index =
+            direction == 0 ? subface_index % 2 : subface_index / 2;
+          return Eval(data.values_within_subface[index],
+                      data.gradients_within_subface[index],
+                      data.hessians_within_subface[index],
+                      data.fe_degree + 1,
+                      data.n_q_points_1d);
+        }
+    }
 
     static void
-    evaluate_in_face(const unsigned int                            n_components,
-                     const MatrixFreeFunctions::ShapeInfo<Number> &data,
-                     Number *                                      values_dofs,
-                     Number *                                      values_quad,
-                     Number *           gradients_quad,
-                     Number *           hessians_quad,
-                     Number *           scratch_data,
-                     const bool         evaluate_val,
-                     const bool         evaluate_grad,
-                     const bool         evaluate_hessian,
-                     const unsigned int subface_index)
+    evaluate_in_face(
+      const unsigned int                                      n_components,
+      const EvaluationFlags::EvaluationFlags                  evaluation_flag,
+      const MatrixFreeFunctions::UnivariateShapeData<Number> &data,
+      Number *                                                values_dofs,
+      Number *                                                values_quad,
+      Number *                                                gradients_quad,
+      Number *                                                hessians_quad,
+      Number *                                                scratch_data,
```

Comment:
Here you keep the raw pointers. I am fine with this - just pointing out.

Reply:
Here I need some shifts in the arrays compared to the other functions (because we call them from places where we operate at one component at a time), so I voted for keeping these functions low-level with individual arguments. We can change this in the future if we feel it makes our lives simpler.

Reply:
OK. Fine with me!